src/parser/parser.h

```diff
@@ -720,12 +720,28 @@ class CVC4_PUBLIC Parser {
    */
   inline size_t scopeLevel() const { return d_symtab->getLevel(); }
 
+  /**
+   * Pushes a scope. All subsequent symbol declarations made are only valid in
+   * this scope, i.e. they are deleted on the next call to popScope.
+   *
+   * The argument bindingLevel is true, the assertion level is set to the
+   * current scope level. This determines which scope assertions are declared
+   * at.
+   */
   inline void pushScope(bool bindingLevel = false) {
     d_symtab->pushScope();
     if(!bindingLevel) {
       d_assertionLevel = scopeLevel();
     }
   }
+  /**
+   * Pushes a scope with new bound variables are bound to types based on the
+   * entries in sortedVarNames. This method returns a bound variable list
+   * containing each of these variables.
+   */
+  Expr pushScopeWithDefs(
```

Comment:
What do you think about doing something like `SynthFunFactory` (#3335), e.g. `DefinitionScope` to make sure that the popping of the scope happens automatically? We can address this in a separate PR.

Reply:
Possibly. There are advantages and disadvantages to this in my opinion. Sometimes it's nice to just see the push/pop scopes in plain sight.